### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by samoht**

### CODE REVIEW

**Comment by samoht** [comment]
This PR is trying to give an overview on how mirage libraries are organised and how Mirage3 works.  /cc @mirage/core let me know if something is not clear enough.

I'll add a few more details in the doc on the mirage+dune PR to explain the new compilation strategy.

File: lib/mirage/mirage.mli [hunk 1]
```diff
@@ -15,19 +15,103 @@
...
+    - A set of {e typed signatures} to make sure these libraries are consistent
+      and can interoperate. As all the library are almost all pure OCaml code,
+      we have defined {e a set of OCaml module types} that encode these
+      conventions in a statically enforcable way. We make no compatibility
+      guarantees at the C level, but compile those on a best-effort basis.
+
+    - Finally, MirageOS is also a {e metaprogramming compiler} that generates
+      OCaml code. It takes as input: the OCaml source code of a program and all
+      of its dependencies, the full description of the deployment target,
+      including configuration values (like the HTTP port to listen too, or the
```
Comment (hannesm) [inline lib/mirage/mirage.mli:39]:
```suggestion
      including configuration values (like the HTTP port to listen on, or the
```

File: lib/mirage/mirage.mli [hunk 2]
```diff
@@ -15,19 +15,103 @@
...
+      private key or the service being deployed). The `mirage`CLI tool uses all
+      of these to {e generate a executable unikernel}: a specialised binary
+      artefact containing only the code what is needed to run on the given
+      deployment platform and no more.
```
Comment (hannesm) [inline lib/mirage/mirage.mli:43]: maybe worth to mention that port and private keys etc. can as well be runtime boot parameters?

File: lib/mirage/mirage.mli [hunk 3]
```diff
@@ -15,19 +15,103 @@
...
+
+    It is possible to write high-level MirageOS applications, such as HTTPS,
+    email or CalDAV servers which can be deployed on very heterogenous and
+    embedded platforms by changing only a few compilation parameters. The
+    supported platforms range from ESP32 micro-controllers, minimal virtual
```
Comment (hannesm) [inline lib/mirage/mirage.mli:48]: I am not sure what you are targeting here -- if it is a description of the MirageOS tool (as released in 3.x series), I'd stay to the supported and implemented targets (I don't think the ESP32 target is merged and released)

File: lib/mirage/mirage.mli [hunk 4]
```diff
@@ -15,19 +15,103 @@
...
+    platform do not have a full POSIX environment; MirageOS does not try to
+    emulate POSIX and focuses on providing a small, well-defined, typed
+    interface with the system components. The nearest equivalent to the MirageOS
+    approach is the WASI (wasi.dev) set of interfaces for WebAssembly.
+
+    {2 Is everything really written in OCaml?}
+
+    While most of the code is written in OCaml, a typed, high-level language
+    with many good safety properties, there are pieces of MirageOS which are
+    still written in unsafe C. These bits can be separated in three categories:
```
Comment (hannesm) [inline lib/mirage/mirage.mli:60]:
```suggestion
    still written in C. These bits can be separated in three categories:
```

File: lib/mirage/mirage.mli (hunk 2 above)
Comment (samoht) [reply lib/mirage/mirage.mli:43]: That paragraph is already quite long, so I'm not sure how to add this succinctly.

File: lib/mirage/mirage.mli (hunk 3 above)
Comment (samoht) [reply lib/mirage/mirage.mli]: fixed

File: lib/mirage/mirage.mli [hunk 5]
```diff
@@ -15,19 +15,103 @@
...
+    {{:https://github.com/ocaml-cross/opam-cross-windows} ocaml-cross} way of
+    doing it is to duplicate {e all} existing opam pacakges by adding a
+    [-windows] suffix to their names and dependencies; this allows normal
+    packages and windows packages can be co-installed in the same opam switch.
+
+    MirageOS 3 solves this by duplicating only the packages defining C bindings
+    and (ab)using
+    {{:https://github.com/ocaml/opam-repository/blob/master/packages/zarith-xen/zarith-xen.1.7/files/mirage-install.sh#L20}
+    ocamlfind predicates} to resolve link time depencencies for C archives and
+    on relying on the now deprecated [--output-obj] option of the OCaml
```
Comment (dinosaure) [inline lib/mirage/mirage.mli:98]: Currently, `--output-obj` is not really deprecated where a discussion still exists on this specific point here: https://github.com/ocaml/ocaml/pull/1351. Then, the `dune` team decided to use `--output-complete-obj` https://github.com/ocaml/dune/pull/23 in the same time.

File: lib/mirage/mirage.mli [hunk 6]
```diff
@@ -15,19 +15,103 @@
...
+      real-world usages. This is the case for low-level bit handling for crypto
+      code (even if we try to make sure allocation is alway handled by the OCaml
+      runtime) as well as arbitrary precision numeric computation (e.g. gmp).
+      Ideally we could image rewriting all of these libraries in OCaml if we had
+      an infinite amount of time in our hands.
+
+    {2 MirageOS as a cross-compilator}
+
+    The MirageOS compiler is basically a cross-compiler, where the host and
+    target toolchain are identical, but with different flags for the C bindings:
```
Comment (hannesm) [inline lib/mirage/mirage.mli:81]: maybe refer to the used pkg-config here? i.e. that CFLAGS are passed from the lowest layers via pkg-config files and invocation.

Comment (dinosaure) [reply lib/mirage/mirage.mli:81]: I think a mention of `pkg-config` and `ocamlfind` as tools needed to correctly handle the cross-compilation is good.

Comment (samoht) [reply lib/mirage/mirage.mli:81]: I've tried to complete that section with more information about pkg-config and ocamlfind predicates.

Comment (dinosaure) [reply lib/mirage/mirage.mli:81]: Seems good for me :+1:.

**Comment by samoht** [comment]
Thanks for the review. More feedback is welcome as subsequent PRs.

**Pull request merged**